scarb/src/internal/serdex.rs

```diff
@@ -1,11 +1,92 @@
-use anyhow::Result;
+use anyhow::{anyhow, Result};
 use camino::{Utf8Path, Utf8PathBuf};
 use serde::{Deserialize, Serialize};
 
 use crate::internal::fsx;
 
-/// Merge two `toml::Value` serializable structs.
-pub fn toml_merge<'de, T, S>(target: &T, source: &S) -> anyhow::Result<T>
+const MERGE_STRATEGY_KEY: &str = "merge-strategy";
+
+#[derive(Debug, Default)]
+pub enum MergeStrategy {
+    #[default]
+    Override,
+    Merge,
+}
+
+impl TryFrom<&str> for MergeStrategy {
+    type Error = anyhow::Error;
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        match value {
+            "override" => Ok(Self::Override),
+            "merge" => Ok(Self::Merge),
+            _ => Err(anyhow!(
+                "invalid merge strategy: {}, must be one of `merge`, `override`",
+                value
+            )),
+        }
+    }
+}
+
+/// Merge `source` into `target` where `source` and `target` are two `toml::Value` serializable
+/// structs.
+/// If `source` and `target` are tables and a specific key exists in both, the conflict will be
+/// resolved as follows:
+/// 1) If value under the conflicting key in either `source` or `target` is not a table, the value
+/// in `target` will be overridden with the value from `source`.
+/// 2) If values under the conflicting key in both `source` and `target` are tables,
+/// the conflict will be resolved with one of two strategies defined by `source`:
+///    a) If `source` have a key `merge-strategy` with value `override`, the value in `target` will
+///     be overridden with the value from `source`.
+///    b) If `source` have a key `merge-strategy` with value `merge`, the value in `target` will be
+///     merged with the value from `source` recursively with this function.
```

Comment:
nit: maybe it would be a good idea to be more clear on the difference between merge/override?
eg it is not clear to me that when I use override, I will need to pass values for all the keys from source (otherwise they are lost)